ddl/column.go

```diff
@@ -0,0 +1,278 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddl
+
+import (
+	"sync/atomic"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/column"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/meta"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/util/errors2"
+)
+
+func (d *ddl) adjustColumnOffset(columns []*model.ColumnInfo, indices []*model.IndexInfo, offset int) {
+	offsetChanged := make(map[int]int)
+	for i := offset; i < len(columns); i++ {
+		offsetChanged[columns[i].Offset] = i
+		columns[i].Offset = i
+	}
+
+	// Update index column offset info.
+	for _, idx := range indices {
+		for _, col := range idx.Columns {
+			newOffset, ok := offsetChanged[col.Offset]
+			if ok {
+				col.Offset = newOffset
+			}
+		}
+	}
+}
+
+func (d *ddl) addColumn(tblInfo *model.TableInfo, spec *AlterSpecification) (*model.ColumnInfo, int, error) {
+	// Check column name duplicate.
+	cols := tblInfo.Columns
+	position := len(cols)
+
+	// Get column position.
+	if spec.Position.Type == ColumnPositionFirst {
+		position = 0
+	} else if spec.Position.Type == ColumnPositionAfter {
+		c := findCol(cols, spec.Position.RelativeColumn)
+		if c == nil {
+			return nil, 0, errors.Errorf("No such column: %v", spec.Column.Name)
+		}
+
+		// Insert position is after the mentioned column.
+		position = c.Offset + 1
+	}
+
+	// TODO: set constraint
+	col, _, err := d.buildColumnAndConstraint(position, spec.Column)
+	if err != nil {
+		return nil, 0, errors.Trace(err)
+	}
+
+	colInfo := &col.ColumnInfo
+	colInfo.State = model.StateNone
+	// To support add column asynchronous, we should mark its offset as the last column.
+	// So that we can use origin column offset to get value from row.
+	colInfo.Offset = len(cols)
+
+	// Insert col into the right place of the column list.
+	newCols := make([]*model.ColumnInfo, 0, len(cols)+1)
+	newCols = append(newCols, cols[:position]...)
+	newCols = append(newCols, colInfo)
+	newCols = append(newCols, cols[position:]...)
+
+	tblInfo.Columns = newCols
+	return colInfo, position, nil
+}
+
+func (d *ddl) onColumnAdd(t *meta.Meta, job *model.Job) error {
```

Comment:
Now we use `onSchemaCreate`, `onIndexCreate` as schema change processing function names, so i think it is better to keep the same.